### PR TITLE
add bytesRead to FileStream

### PR DESCRIPTION
**Pull request by Uzlopak**
Inspired by https://github.com/mscdex/busboy/pull/174 by @ahuigo

FileStream will be similar to FS.ReadStream and provide the attribute bytesRead. 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/types/multipart.js [hunk 1]
```diff
@@ -199,6 +199,8 @@ function Multipart (boy, cfg) {
             file.truncated = true
             part.removeAllListeners('data')
           } else if (!file.push(data)) { self._pause = true }
+
+          file.bytesRead = nsize > fileSizeLimit ? fileSizeLimit : nsize
```
Comment (kibertoad) [inline lib/types/multipart.js:203]: does this cause any noticeable perf impact?

Comment (Uzlopak) [reply lib/types/multipart.js:203]: with the current busboy benchmarks, no.

But we can wait till you have created more reliable benchmarks.  :)

Comment (Uzlopak) [reply lib/types/multipart.js:203]: @kibertoad  Can you please check if this has some relevant performance impact?

Comment (kibertoad) [reply lib/types/multipart.js:203]: will do!

**Comment by kibertoad** [comment]
Before: Mean time for fastify-busboy is 291910.52376131905 nanoseconds
After: Mean time for fastify-busboy is 295153.5812366501 nanoseconds

I've ran it on medium preset, twice, results are fairly consistent. There is approximately 1% loss of performance.

Of course, it is still faster than the original:
Mean time for busboy is 366253.787385089 nanoseconds

@Uzlopak Thoughts?

**Comment by Uzlopak** [comment]
On second thought I think this is not the best solution. I reread the code and I think we just need to assign nsize in that branch and the branch above when we truncate the file. There is I think no need for the ternary operation. Will check that later.

**Comment by Uzlopak** [comment]
Is the mathemal changes made by me correct?

**Comment by kibertoad** [comment]
Will be back in few hours and will check!

**Comment by kibertoad** [comment]
wonder if we can have interim state unit test for this

**Comment by Uzlopak** [comment]
I actually find it prone to a race condition that I first emit the limit event and later set the bytesRead attribute. That's why I moved the truncated attribute change above the emit. 

Should we set the bytesRead above the emit? Additional to the bytesRead below the if block. 

In worst case bytesRead is set twice it limit is hit, which is imho acceptable.

Wdyt?

**Comment by Uzlopak** [comment]
made it less race condition prone

File: lib/types/multipart.js [hunk 2]
```diff
@@ -195,12 +195,17 @@ function Multipart (boy, cfg) {
 
         onData = function (data) {
           if ((nsize += data.length) > fileSizeLimit) {
-            const extralen = (fileSizeLimit - (nsize - data.length))
+            const extralen = nsize - fileSizeLimit
```
Comment (kibertoad) [inline lib/types/multipart.js:198]: can you explain your train of thought behind this change? This is clearly very different to the original math, but I assume there is method to it?

Comment (Uzlopak) [reply lib/types/multipart.js]: Ok, this was quick math. But that should be fixed now

**Review by kibertoad — APPROVED** [review]

**Pull request merged**